source/server/http/stats_handler.cc

```diff
@@ -190,52 +190,99 @@ uint64_t PrometheusStatsFormatter::statsAsPrometheus(
+   * All lines for a given metric must be provided as one single group, with the optional HELP and
+   * TYPE lines first (in no particular order). Beyond that, reproducible sorting in repeated
+   * expositions is preferred but not required, i.e. do not sort if the computational cost is
+   * prohibitive.
+   */
+
+  /**
+   * Processes a metric type (counter, gauge, histogram) by generating all output lines, sorting
+   * them by tag-extracted metric name, and then outputting them in the correct sorted order into
+   * response.
+   *
+   * @param metrics A vector of pointers to a metric type.
+   * @param generate_output A std::function<std::string(const MetricType& metric, const std::string&
+   *        prefixedTagExtractedName)> which returns the output text for this metric.
+   */
+  auto process_type = [&](const auto& metrics, const auto& generate_output,
+                          absl::string_view type) -> uint64_t {
+    using MetricType = typename std::remove_reference<decltype(metrics)>::type::value_type;
+
+    struct MetricLessThan {
+      bool operator()(const MetricType& a, const MetricType& b) const {
+        ASSERT(&a->constSymbolTable() == &b->constSymbolTable());
+        return a->constSymbolTable().lessThan(a->statName(), b->statName());
+      }
+    };
 
-  for (const auto& gauge : gauges) {
-    if (!shouldShowMetric(*gauge, used_only, regex)) {
-      continue;
-    }
+    // This is a sorted collection to satisfy the "preferred" ordering from the prometheus
+    // spec: metrics will be sorted by their tags' textual representation, which will be consistent
+    // across calls.
+    using MetricTypeSortedCollection = std::set<MetricType, MetricLessThan>;
 
-    const std::string tags = formattedTags(gauge->tags());
-    const std::string metric_name = metricName(gauge->tagExtractedName());
-    if (metric_type_tracker.find(metric_name) == metric_type_tracker.end()) {
-      metric_type_tracker.insert(metric_name);
-      response.add(fmt::format("# TYPE {0} gauge\n", metric_name));
+    // Return early to avoid crashing when getting the symbol table from the first metric.
+    if (metrics.empty()) {
+      return 0;
     }
-    response.add(fmt::format("{0}{{{1}}} {2}\n", metric_name, tags, gauge->value()));
-  }
 
-  for (const auto& histogram : histograms) {
-    if (!shouldShowMetric(*histogram, used_only, regex)) {
-      continue;
-    }
+    // There should only be one symbol table for all of the stats in the admin
+    // interface. If this assumption changes, the name comparisons in this function
+    // will have to change to compare to convert all StatNames to strings before
+    // comparison.
+    const Stats::SymbolTable& global_symbol_table = metrics.front()->constSymbolTable();
+
+    // Sorted collection of metrics sorted by their tagExtractedName, to satisfy the requirements
+    // of the exposition format.
+    std::map<Stats::StatName, MetricTypeSortedCollection, Stats::StatNameLessThan> groups(
```

Comment:
this looks much better, thank you! WDYT of using a std::vector and calling std::sort -- my cargo-culted opinion is that that will be faster and more memory-efficient (by a constant factor) than using std::map as a sorter.

Ditto for the std::set in MetricTypeSortedCollection; once you have the vector populated you can just sort it.

Reply:
That would be fine for the inner std::set. For the outer std::map, we're doing a lot of lookups by key. I could use a flat_hash_map and then order it at the end, but that sounds painful; I don't think there's a better way to do it than to first make an unordered-copy into a vector and then sort the vector.

Reply:
That makes sense, sgtm.